add Colyseus + PlayCanvas tutorial

endel opened the pull request:
Hi everybody! I hope you are well and having a great start in 2022!

Me and @lpsandaruwan have been working on this tutorial project using Colyseus as an alternative for doing real-time networking on PlayCanvas. We hope this could fit well on the tutorials page. 🙏 

Please let us know if you have any feedback, we'll gladly make any changes ☺️

---

PlayCanvas project: https://playcanvas.com/project/859259/overview/colyseus-demo

---

![image](https://user-images.githubusercontent.com/16639049/150157237-ecc55e8b-3105-48e3-9fee-3a06bc579328.png)

jpauloruschel commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,403 @@
+
+Now, from a new PlayCanvas Script, let's instantiate our `Colyseus.Client` instance. ([see "Creating new scripts"](https://developer.playcanvas.com/en/user-manual/scripting/creating-new/))
+
+You can attach this script to any object in your scene, such as the Camera.
```
I think we can help diffuse some project best practices and instead suggest to add this script to a new empty entity (named something like "NetworkManager").

jpauloruschel commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,403 @@
+
+---
+
+So, let's define our Room State in the Server.
```
(super minor): when following these steps, my attention was at this point on the PlayCanvas editor. Might be worth it to make it clearer to "go back to editing the Server code".

yaustar replied:
```suggestion
Let's define our Room State in the Server.
```

jpauloruschel commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,403 @@
+        const player = new Player();
+
+        // place Player at a random position
+        player.x = Math.random() * 7.2 - 3.6;
```
These numbers seem arbitrary?

yaustar commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,403 @@
+## Materials
+
+- [PlayCanvas Project (Client-side)](https://playcanvas.com/project/859259/overview/colyseus-demo)
+- [Colyseus TypeScript Project (Server-side)](https://github.com/lpsandaruwan/colyseus-playcanvas-server/) ([Glitch Server](https://glitch.com/edit/#!/tutorial-playcanvas-server))
```
I was a bit confused by the Glitch server. It looks like it can be a separate option to host the server for Colyseus. 

Maybe this should be a separate section at the end of this doc about hosting the server?

yaustar commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,403 @@
+
+# Creating the Server
+
+We will be making a basic server for keeping player states. Changes will be synchronized with clients accordingly.
```
```suggestion
We will be making a basic server, hosted locally on your computer for keeping player states. Changes will be synchronized with clients accordingly.
```

yaustar commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
````diff
@@ -0,0 +1,403 @@
+}
+```
+
+> Note that we're using the local `ws://localhost:2567` endpoint here. You need to [deploy your server](https://docs.colyseus.io/arena/getting-started/create-application/) to the public internet in order to play with others online.
````
Maybe reference using Glitch as an option here or link to the section?

yaustar commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,403 @@
+
+Notice that `my_room` is the default room identifier set by the barebones Colyseus server. You can and should change this identifier in the `arena.config.ts` file.
+
+You must be seeing the following message in your server logs, which means a client successfully joined the room!
```
```suggestion
You will be seeing the following message in your server logs, which means a client successfully joined the room!
```

yaustar commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,403 @@
+
+# Room State and Schema
+
+In Colyseus, we define shared data through the `Schema` structures.
+
+The state synchronization loop looks like this:
+
+1. State mutations are synchronized automatically from Server → Clients
+2. Clients, by attaching callbacks to their local _read-only_ `Schema` structures, can observe for state mutations and react to it.
+3. Clients can send arbitrary messages to the server - which decides what to do with it - and may mutate the state (Go back to step **1.**)
```
There are several technical terms used here that may need explaining:

`Schema` and `mutations`

yaustar commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,403 @@
+
+Make sure to uncheck the `"Enabled"` property. We will not have any Player instances enabled until we have active connections with the server.
+
+![Player](/images/tutorials/multiplayer-colyseus/player.png)
```
Maybe highlight the change needed in the image?

yaustar commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,403 @@
+
+When playing the Scene, you should see a message in the browser's console whenever a new client joins the room.
+
+Now, for the visual representation, we need to clone the "Player" object, and keep a local reference to the cloned object based on their `sessionId`, so we can operate on them later:
```
```suggestion
For the visual representation, we need to clone the "Player" object, and keep a local reference to the cloned object based on their `sessionId`, so we can operate on them later:
```

yaustar commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,403 @@
+
+# More
+
+We hope you found this tutorial useful, if you'd like to learn more about Colyseus please have a look at the [Colyseus documentation](https://docs.colyseus.io/), and join the [Discord community](https://discord.gg/RY8rRS7).
```
```suggestion
We hope you found this tutorial useful, if you'd like to learn more about Colyseus please have a look at the [Colyseus documentation](https://docs.colyseus.io/), and join the [Colyseus Discord community](https://discord.gg/RY8rRS7).
```

endel commented:
Thank you so much for the thorough feedback @yaustar and @jpauloruschel 🙏 

Really glad you're open to receiving contributions here! I've just updated the PR addressing all your suggestions!

Please let me know if you find anything else that needs clarification and/or better explanation.

This tutorial is using the upcoming (not officially released yet) version 0.15 of Colyseus, I hope you don't mind if I submit another pull request a few days from now to update the version from `colyseus.js@^0.15.0-preview.2` to `colyseus.js@^0.15.0` 😇 

Cheers!

Maksims commented:
In tutorial there are places where "not the best practices" are used, e.g. allocating vectors every update. This should be avoided.
Refer to User Manual docs with performance and best practices for scripting.

willeastcott commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,406 @@
+  entity.enabled = true;
+
+  // set position based on server data
+  entity.setPosition(new pc.Vec3(player.x, player.y, player.z));
```
```suggestion
entity.setPosition(player.x, player.y, player.z);
```

willeastcott commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,406 @@
+  entity.setPosition(new pc.Vec3(player.x, player.y, player.z));
+
+  // add clone to the Scene
+  playerEntityToClone.getParent().addChild(entity);
```
```suggestion
  playerEntityToClone.parent.addChild(entity);
```

willeastcott commented on content/en/tutorials/real-time-multiplayer-colyseus.md:
```diff
@@ -0,0 +1,406 @@
+this.room.state.players.onAdd((player, sessionId) => {
+  // ...
+  player.onChange(() => {
+    const newPosition = new pc.Vec3(player.x, player.y, player.z);
+    this.playerEntities[sessionId].setPosition(newPosition);
```
```suggestion
    this.playerEntities[sessionId].setPosition(player.x, player.y, player.z);
```

endel replied:
Lovely! Thank you @willeastcott!

willeastcott commented:
@Maksims - I've tried to address your comments with some suggested code changes. @endel has merged them. Hopefully, you're good with these fixes.

yaustar commented:
Looks like the project code has also been updated too :)

If you are good @Maksims, I will look to merge and publish this.

@endel What we would like to do is to transfer the project to the PlayCanvas ORG where you would still have read + write access.

If changes are needed for the tutorial (eg you would like to update the tutorial for a new feature), we would ask that the project changes be made in a branch on PlayCanvas for review at the same time as a PR on this repo :) 

Is that okay with you?

endel commented:
Hi @yaustar, awesome, sure, that's no problem! Would you like to have the server transferred to PlayCanvas on GitHub as well or we can keep that one?

Just sent the transfer request for the PlayCanvas project. If possible can we keep me (@endel) and @lpsandaruwan with read/write access there?

Cheers! It was nice interacting with you all here! 👏 👏

yaustar commented:
> we can keep that one? (https://github.com/colyseus/tutorial-playcanvas-server)

I say keep that on your side I think for the moment

> Just sent the transfer request for the PlayCanvas project. If possible can we keep me (@endel) and @lpsandaruwan with read/write access there?

Done, thanks!

yaustar commented:
NB: We updated the tutorial project to use a different font as the one that was in the project was hard to read on the tutorial page 😅
